[SRVCOM-2368] Serverless multitenant istio docs (#68618)
openshift/openshift-docs: modules/serverless-ossm-installing-and-configuring-openshift-service-mesh.adoc

:_content-type: PROCEDURE
[id="serverless-ossm-installing-and-configuring-openshift-service-mesh_{context}"]
= Installing and configuring {SMProductShortName}

To integrate {ServerlessProductShortName} with {SMProductShortName}, you need to install {SMProductShortName} with a specific configuration.

.Procedure

. Create a `ServiceMeshControlPlane` resource in the `istio-system` namespace with the following configuration:
+
[IMPORTANT]
====

> If you have an existing `ServiceMeshControlPlane`, make sure that you have the same configuration applied.

Reviewer: ```suggestion
If you have an existing `ServiceMeshControlPlane` object, make sure that you have the same configuration applied.
```

====
+
[source,yaml]
----
apiVersion: maistra.io/v2
kind: ServiceMeshControlPlane
metadata:
  name: basic
  namespace: istio-system
spec:
  profiles:
  - default
  security:
    dataPlane:
      mtls: true <1>
  techPreview:
    meshConfig:
      defaultConfig:
        terminationDrainDuration: 35s <2>
  gateways:
    ingress:
      service:
        metadata:
          labels:
            knative: ingressgateway <3>
  proxy:
    networking:
      trafficControl:
        inbound:
          excludedPorts: <4>
          - 8444 # metrics
          - 8022 # serving: wait-for-drain k8s pre-stop hook
----
<1> Enforce strict mTLS in the mesh. Only calls using a valid client certificate are allowed.
<2> {ServerlessProductShortName} has a graceful termination for Knative Services of 30 seconds. `istio-proxy` needs to have a longer termination duration to make sure no requests are dropped.
<3> Define a specific selector for the ingress gateway to target only the Knative gateway.

> <4> Those ports are called by Kubernetes and cluster monitoring which are not part of the mesh and cannot be called using mTLS. Hence, these ports are excluded from the mesh.

Reviewer 2: Rewrite as: "These ports are called by Kubernetes and cluster monitoring, which are not part of the mesh and cannot be called using mTLS. Therefore, these ports are excluded from the mesh."

> . Add the namespaces that you would like to integrate with {SMProductShortName} to the `ServiceMeshMemberRoll` object as members:

Reviewer: For a follow-up: I noticed earlier there were a lot of "part of the `ServiceMeshMemberRoll`" or "member of the `ServiceMeshMemberRoll`" without an "object" on the end, but I didn't call out as I wasn't sure of the implication. But it looks like here that you've got "object" on the end, so suggest going back to the others and adding for consistency as well.

+
.Example `servicemesh-member-roll.yaml` configuration file
[source,yaml]
----
apiVersion: maistra.io/v1
kind: ServiceMeshMemberRoll
metadata:
  name: default
  namespace: istio-system
spec:
  members: <1>
    - knative-serving
    - knative-eventing
    - your-OpenShift-projects
----
<1> A list of namespaces to be integrated with {SMProductShortName}.
+
[IMPORTANT]
====
This list of namespaces must include the `knative-serving` and `knative-eventing` namespaces.
====

. Apply the `ServiceMeshMemberRoll` resource:
+
[source,terminal]
----
$ oc apply -f servicemesh-member-roll.yaml
----

> . Create the necessary gateways so that {SMProductShortName} can accept traffic. The following example uses `knative-local-gateway` object using ISTIO_MUTUAL (mTLS):

Reviewer: Markup "ISTIO_MUTUAL". Also seems like a few words might be missing. Suggestion, if still technically accurate:
```suggestion
. Create the necessary gateways so that {SMProductShortName} can accept traffic. The following example uses the `knative-local-gateway` object with the `ISTIO_MUTUAL` mode (mTLS):
```

+
.Example `istio-knative-gateways.yaml` configuration file
[source,yaml]
----
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
  name: knative-ingress-gateway
  namespace: knative-serving
spec:
  selector:
    knative: ingressgateway
  servers:
    - port:
        number: 443
        name: https
        protocol: HTTPS
      hosts:
        - "*"
      tls:
        mode: SIMPLE

>         credentialName: <wildcard_certs> <1>

Reviewer 3: I guess it is not clear how to generate the wildcard certificate. We remove some info about the process earlier [in this PR](https://github.com/openshift/openshift-docs/pull/68618/files#diff-32ad31e2767596655f95c2359bda26fa82b65db166ff11126f9ae567f20b61a5L23) and I can't find it anymore. Was it moved somewhere?

Reviewer 4: Good point, we should probably keep that block (no I don't think we moved it somewhere, I just did not add it in the draft docs).

---
apiVersion: networking.istio.io/v1alpha3
kind: Gateway
metadata:
 name: knative-local-gateway
 namespace: knative-serving
spec:
 selector:
   knative: ingressgateway
 servers:
   - port:
       number: 8081
       name: https
       protocol: HTTPS <2>
     tls:
       mode: ISTIO_MUTUAL <2>
     hosts:
       - "*"
---
apiVersion: v1
kind: Service
metadata:
 name: knative-local-gateway
 namespace: istio-system
 labels:
   experimental.istio.io/disable-gateway-port-translation: "true"
spec:
 type: ClusterIP
 selector:
   istio: ingressgateway
 ports:
   - name: http2
     port: 80
     targetPort: 8081
----
<1> Name of the secret containing the wildcard certificate.

> <2> The `knative-local-gateway` serves HTTPS traffic and expects all clients to send requests using mTLS. This means that only traffic coming from within {SMProductShortName} is possible. Workloads from outside the {SMProductShortName} must use the external domain via OpenShift Routing.

Reviewer: ```suggestion
<2> The `knative-local-gateway` object serves HTTPS traffic and expects all clients to send requests using mTLS. This means that only traffic coming from within {SMProductShortName} is possible. Workloads from outside the {SMProductShortName} must use the external domain via OpenShift Routing.
```

----
<1> Name of the secret containing the wildcard certificate.
<2> The `knative-local-gateway` serves HTTPS traffic and expects all clients to send requests using mTLS. This means that only traffic coming from within {SMProductShortName} is possible. Workloads from outside the {SMProductShortName} must use the external domain via OpenShift Routing.
+
[NOTE]
====

> You must configure the wildcard certificate to match the domain of your {ocp-product-title} cluster. For example, if your {ocp-product-title} console address is `https://console-openshift-console.apps.openshift.example.com`, you must configure the wildcard certificate so that the domain is `*.apps.openshift.example.com`.

Reviewer 3: @rh-max Thanks for this. We probably still want to put back the note that we removed:
```
For more information about configuring wildcard certificates, see the following topic about _Creating a certificate to encrypt incoming external traffic_.
 ...
 include::modules/serverless-ossm-external-certs.adoc[leveloffset=+1]
```
And also put back the file serverless-ossm-external-certs.adoc which you're removing in this PR.
Having this example is quite important. Generating the certificates needs a lot of specific parameters so it's hart to figure it out without any example.

Author: @mgencur Thanks. I have removed this particular instance of the note, because I re-added the note in the "Prerequisites" section at the top of the page (and it refers to the "Creating a certificate to encrypt incoming external traffic" section, which follows right after).
